### PR TITLE
frontend: use `mimirpb.Sample` custom JSON encoder func

### PR DESCRIPTION
**Pull request by ortuman**
Signed-off-by: Miguel Ángel Ortuño <ortuman@gmail.com>

#### What this PR does

Use custom `mimir.Sample` JSON serializer hook to minimize allocation when encoding prometheus response in frontend component. This is functionally equivalent and about `~50%` faster in my testing.

This behavior has been [inspired/copied](https://github.com/prometheus/prometheus/blob/main/web/api/v1/api.go#L205) from Prometheus, but applying it to `mimir.Sample` instead of `promql.Point`. 

The following commands can be run to verify the benchmark results:

*Previous*

Comment [this line](https://github.com/grafana/mimir/blob/ortuman/improve-frontend-response-encoding/pkg/frontend/querymiddleware/codec.go#L56) to avoid registering the hook and run the following command: 

```sh
go test -cpu 1 -count=10 -run=XXX -benchmem -bench=BenchmarkPrometheusCodec_EncodeResponse ./pkg/frontend/querymiddleware | tee old.txt
```

*Current*

Uncomment the previous line and run:

```sh
go test -cpu 1 -count=10 -run=XXX -benchmem -bench=BenchmarkPrometheusCodec_EncodeResponse ./pkg/frontend/querymiddleware | tee new.txt
```

Comparison:

```
benchstat old.txt new.txt
name                            old time/op    new time/op    delta
PrometheusCodec_EncodeResponse     190ms ± 1%      93ms ± 1%  -51.16%  (p=0.000 n=9+8)

name                            old alloc/op   new alloc/op   delta
PrometheusCodec_EncodeResponse     201MB ± 2%     100MB ±23%  -50.08%  (p=0.000 n=8+10)

name                            old allocs/op  new allocs/op  delta
PrometheusCodec_EncodeResponse     2.01M ± 0%     0.01M ± 0%  -99.50%  (p=0.000 n=9+10)
```

#### Which issue(s) this PR fixes or relates to

Fixes N/A

#### Checklist

- [X] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/frontend/querymiddleware/codec.go
Comment (pracucci) [inline pkg/frontend/querymiddleware/codec.go:55]: I don't think the custom marshalling code you've added belongs to this file. I would move it to `pkg/mimirpb/` package instead.

File: pkg/frontend/querymiddleware/codec.go
Comment (pracucci) [inline pkg/frontend/querymiddleware/codec.go:463]: Would be nice if you could try to expose `marshalTimestamp()` and `marshalValue()`  in Prometheus, and move them to an util package, so that we could just import that instead of copying it. Given how tricky are these functions, I would prefer to keep them aligned with Prometheus (with the hope Prometheus maintainers don't have objections on the refactoring).

**Review by pracucci — COMMENTED** [review]
Good job! I don't see why we shouldn't do it, given Prometheus does it. I left a couple of comments.

File: pkg/frontend/querymiddleware/codec_test.go
Comment (replay) [inline pkg/frontend/querymiddleware/codec_test.go:680]: I'm trying to understand where this change comes from? That's completely unrelated to the marshaling function, right?

**Review by replay — APPROVED** [review]
Looks great, thanks!

File: pkg/frontend/querymiddleware/codec_test.go
Comment (ortuman) [reply pkg/frontend/querymiddleware/codec_test.go:680]: This previous `vector` result type value prevented `BenchmarkPrometheusCodec_DecodeResponse` from being run. 

The following error was prompted:

```
BenchmarkPrometheusCodec_DecodeResponse
    codec_test.go:618: 
        	Error Trace:	/Users/ortuman/Projects/mimir/pkg/frontend/querymiddleware/codec_test.go:618
        	Error:      	Received unexpected error:
        	            	querymiddleware.PrometheusResponse.Data: can't marshal prometheus result type "value"
        	Test:       	BenchmarkPrometheusCodec_DecodeResponse
--- FAIL: BenchmarkPrometheusCodec_DecodeResponse
FAIL
```

File: pkg/frontend/querymiddleware/codec.go
Comment (ortuman) [reply pkg/frontend/querymiddleware/codec.go]: done @ da7291e893b606b8c5ef33fddb3c06a339df702d

File: pkg/frontend/querymiddleware/codec.go
Comment (ortuman) [reply pkg/frontend/querymiddleware/codec.go]: Followed your suggestion and created a Prometheus PR to export point marshaling functions. https://github.com/prometheus/prometheus/pull/11323

I guess we could keep these copied functions at `mimirpb` and adapt mimir code later. But if you consider its better to wait until the upstream change gets merged and leave this PR on-hold, I'm ok with that.  👍

File: pkg/frontend/querymiddleware/codec_test.go
Comment (replay) [reply pkg/frontend/querymiddleware/codec_test.go:680]: thanks for explaining

File: pkg/frontend/querymiddleware/codec.go
Comment (colega) [reply pkg/frontend/querymiddleware/codec.go]: I would move forward with this, and call Prometheus' exported functions once they're merged.

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:110]: Can you move it up, under "main", please?

File: pkg/mimirpb/compat.go
Comment (pracucci) [inline pkg/mimirpb/compat.go:241]: [nit] Don't need this. You already have `sample := (*Sample)(ptr)` above.

**Review by pracucci — APPROVED** [review]
LGTM!

File: CHANGELOG.md
Comment (ortuman) [reply CHANGELOG.md]: done @ 7e3ae47e6a6787189f206323179e2fc8a421686d

File: pkg/mimirpb/compat.go
Comment (ortuman) [reply pkg/mimirpb/compat.go]: done @ 7e3ae47e6a6787189f206323179e2fc8a421686d

File: CHANGELOG.md
Comment (pracucci) [inline CHANGELOG.md:26]:
```suggestion
* [ENHANCEMENT] Query-frontend: improved Prometheus reponse JSON encoding performance. #2450
```

**Pull request merged**
